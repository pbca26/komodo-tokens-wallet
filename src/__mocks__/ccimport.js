// IMPORTANT: cclib must be compiled with nodejs target
//            otherwise Jest won't be able to load it
import * as ccimp from '../../cryptoconditions-js-tests/pkg/cryptoconditions.js';
window.ccimp = ccimp;