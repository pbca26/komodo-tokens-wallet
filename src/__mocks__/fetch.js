import mockData from './blockchain.json';

const mockFetch = jest.fn((url) => {
  console.log('fetch url', url);

  return Promise.resolve({
    json: () => {
      if (url.indexOf('status?q=getInfo') > -1) {
        return mockData.getInfo;
      } if (url.indexOf('tx/send') > -1) {
        return {txid: '1111111111111111111111111111111111111111111111111111111111111111'};
      } if (url.indexOf('rawtx') > -1) {
        if (url.indexOf('/')) {
          const temp = url.split('/');
          return mockData.rawtx[temp[1]];
        }

        return mockData.rawtx['890a375c18cbc420d3293dbd661f557caf5e4ca43c770cf479bcc9cfe14b9cbd'];
      } else if (url.indexOf('addr/undefined/utxo') > -1 || url.indexOf('addr/RRF6ocq94kLpaH6wPdZjTqQa9mAmT1GToc/utxo') > -1) {
        return mockData.utxo;
      } else if (url === 'tokens') {
        return mockData.tokens;
      } else if (url.indexOf('tokens/balance') > -1) {
        return mockData.balance;
      } else if (url.indexOf('transactions?address=undefined') > -1 || url.indexOf('transactions?address=CaopajuemreFwEAfbywvFi8oFyrJPNkDs1') > -1) {
        return mockData.transactions;
      } else if (url.indexOf('tokens/orderbook') > -1) {
        return mockData.orderbook;
      } else if (url.indexOf('tokens/createtx') > -1) {
        if (url.indexOf('tokens/createtx?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&amount=10000') > -1) {
          return mockData.cancelAsk.createtx;
        } else if (url.indexOf('tokens/createtx?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&amount=11000' > -1)) {
          return mockData.sellToken.createtx;
        }
        return mockData.createtx;
      } else if (url.indexOf('tokens/addccinputs') > -1) {
        if (url.indexOf('tokens/addccinputs?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&tokenid=7e2b623cb57b44b8dc5d3a3cc36c20d08f66023a85fa2a4490a8fb8783d38347&amount=1' > -1)) {
          return mockData.sellToken.addccinputs;
        }
        return mockData.createToken.fixedSupplyToken.chainData;
      } else if (url.indexOf('tokens/utxo') > -1) {
        return mockData.ccUtxo;
      } else if (url.indexOf('tokens/transactionsmany?txid1') > -1) {
        if (url.indexOf('tokens/transactionsmany?txid1=b8ceac7c2130b9a44fd11a1b67f4fa7b7a908b9f9a58e4f0024acd3b486b7a90&txid2=572663bb71d9822cf70d60ed4d180300f61bf8790dc5c2023d1888f1b8e8c254') > -1) {
          return mockData.cancelAsk.transactionsmany;
        }
        return mockData.transactionsMany;
      }
    },
    headers: {
      get: () => {
        return 'application/json';
      },
    },
    ok: true,
  });
});

export default mockFetch;