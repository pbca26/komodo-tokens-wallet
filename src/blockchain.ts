// TODO: rewrite as class
let explorerUrl: string, _fetch = typeof window === 'undefined' ? null : fetch, _headers = typeof window === 'undefined' ? null : Headers;

const setExplorerUrl = (name?: string) => {
  explorerUrl = name;
};

const setFetch = (fetchRef: any, headers: any) => {
  _fetch = fetchRef;
  _headers = headers;
};

const get = async (endpoint: string, postData?: any) => {
  const opts: any = {};

  if (postData) {
    opts.body = JSON.stringify(postData);
    opts.headers = new _headers();
    opts.headers.append('Content-Type', 'application/json');
    opts.headers.append('Content-Length', opts.body.length);
    opts.method = 'POST';
  }

  const response = await _fetch(`${explorerUrl}${endpoint}`, opts);
  const isJson = response.headers.get('Content-Type').includes('application/json');

  const body = isJson ? await response.json() : await response.text();

  if (!response.ok) {
    throw new Error(body);
  }

  return body;
};

const getAddress = (address: string) => get(`addr/${address}/?noTxList=1`);

const getAddressHistory = (address: string) => get(`/txs?address=${address}`);

const getHistory = (addresses: string[]) => get(`addrs/txs`, {addrs: addresses.join(',')});

//const getUtxos = (address: string) => get(`addrs/utxo`, {addrs: addresses.join(',')});
const getNormalUtxos = (address: string) => get(`addr/${address}/utxo`);

const getTransaction = (txid: string) => get(`tx/${txid}`);

const getRawTransaction = (txid: string) => get(`rawtx/${txid}`);

const getBestBlockHash = () => get('status?q=getBestBlockHash');

const getBlock = (blockHash: string) => get(`block/${blockHash}`);

const getTipTime = async () => {
  const {bestblockhash} = await getBestBlockHash();
  const block = await getBlock(bestblockhash);

  return block.time;
};

const broadcast = (transaction: string) => get('tx/send', {rawtx: transaction});

const getInfo = async () => {
  try {
    const response = await fetch(`${explorerUrl}/status?q=getInfo`);
    const isJson = response.headers.get('Content-Type').includes('application/json');

    const body = isJson ? await response.json() : await response.text();
    
    if (!response.ok) {
      throw new Error(body);
    }

    return body;
  } catch (e) {
    return null;
  }
};

const tokenBalance = (address: string) => get(`tokens/balance?address=${address}`);
const tokenTransactions = (address: string) => get(`tokens/transactions?address=${address}`);
const tokenList = (cctxid?: string | string[]) => cctxid ? get('tokens', {cctxid}) : get('tokens');
const tokenListAll = () => get('tokens?pageNum=all');
const tokenOrderbook = (address?: string) => address ? get(`tokens/orderbook?address=${address}`) : get('tokens/orderbook');
const addCCInputs = (tokenid: string, pubkey: string, amount: number) => get(`tokens/addccinputs?pubkey=${pubkey}&tokenid=${tokenid}&amount=${amount}`);
const createCCTx = (amount: number | string, pubkey: string) => get(`tokens/createtx?pubkey=${pubkey}&amount=${amount}`);
const tokenUtxos = (address: string, tokenid: string, raw = false) => get(`tokens/utxo?address=${address}&cctxid=${tokenid}&raw=${raw}`)
const tokenTransactionsMany = (txid1: string, txid2: string) => get(`tokens/transactionsmany?txid1=${txid1}&txid2=${txid2}`);
const raddressToCIndexKeys = (address: string) => get(`tokens/convertAddress?address=${address}`);

const blockchain = {
  get,
  getInfo,
  getAddress,
  getAddressHistory,
  getHistory,
  getTransaction,
  getRawTransaction,
  getBestBlockHash,
  getBlock,
  getTipTime,
  broadcast,
  setExplorerUrl,
  setFetch,
  tokenBalance,
  tokenTransactions,
  tokenList,
  tokenListAll,
  getNormalUtxos,
  addCCInputs,
  createCCTx,
  tokenUtxos,
  tokenTransactionsMany,
  tokenOrderbook,
  raddressToCIndexKeys,
};

export default blockchain;