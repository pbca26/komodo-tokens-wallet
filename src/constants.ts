import {IChains} from './types';

export const chains: IChains = {
  TOKENSV2: {
    ccLibVersion: 1,
    faucetURL: 'https://www.atomicexplorer.com/#/faucet/tokensv2/',
    txBuilderApi: 'utxoSelect', // default|insight|utxoSelect
    explorerUrl: 'http://explorer.komodoplatform.com:20000/tokens',
    explorerApiUrl: 'https://explorer.komodoplatform.com:10000/tokensv2/api/',
    enabled: true,
  },
  TKLTEST: {
    ccLibVersion: 2,
    ccIndex: true,
    explorerApiVersion: 2,
    faucetURL: 'https://www.atomicexplorer.com/#/faucet/tkltest/',
    txBuilderApi: 'utxoSelect', // default|insight|utxoSelect
    explorerUrl: 'http://explorer.komodoplatform.com:20000/tokens',
    explorerApiUrl: 'https://explorer.komodoplatform.com:10000/tkltest/api/',
    enabled: true,
  },
  TKLTEST2: {
    ccLibVersion: 2,
    ccIndex: true,
    explorerApiVersion: 2,
    faucetURL: 'https://www.atomicexplorer.com/#/faucet/tkltest2/',
    txBuilderApi: 'utxoSelect', // default|insight|utxoSelect
    explorerUrl: 'http://explorer.komodoplatform.com:20000/tokens',
    explorerApiUrl: 'https://explorer.komodoplatform.com:10000/tkltest2/api/',
    enabled: true,
  },
  TKLTEST2R: {
    ccLibVersion: 2,
    ccIndex: true,
    explorerApiVersion: 2,
    faucetURL: 'https://www.atomicexplorer.com/#/faucet/tkltest2r/',
    txBuilderApi: 'utxoSelect', // default|insight|utxoSelect
    explorerUrl: 'http://explorer.komodoplatform.com:20000/tokens',
    explorerApiUrl: 'https://explorer.komodoplatform.com:10000/tkltest2r/api/',
    enabled: true,
    isRaddress: true,
  },
  TOKEL: {
    ccLibVersion: 2,
    ccIndex: true,
    txBuilderApi: 'utxoSelect', // default|insight|utxoSelect
    explorerUrl: 'http://explorer.komodoplatform.com:20000/tokens',
    explorerApiUrl: 'https://explorer.komodoplatform.com:10000/tokel/api/',
    explorerApiVersion: 2,
    enabled: window.location.href.indexOf('enable-tokel') > -1 ? true : false,
  },
};

export const nftDataTypes: { [key: string]: string } = {
  plain: 'Plain text or JSON',
  tokel: 'Tokel Standart',
};

export const orderType: Array<{ [char: string]: string }> = [{
  value: 'all',
  title: 'All orders',
}, {
  value: 'my',
  title: 'My orders',
}];

export const orderDirection: Array<{ [char: string]: string }> = [{
  value: 'all',
  title: 'All directions',
}, {
  value: 'sell',
  title: 'Sell orders',
}, {
  value: 'buy',
  title: 'Buy orders',
}];