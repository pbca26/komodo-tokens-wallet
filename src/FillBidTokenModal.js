import React, {useState, useEffect} from 'react';
import Modal from './Modal';
import TokensLib from './cclib-import';
import Blockchain from './blockchain';
import {chains} from './constants';
import {utxoSelectCC, utxoSelectNormal} from './utxo-select';
import {toSats} from './math';
import {getMaxSpendNormalUtxos} from './math';
import writeLog from './log';
import devVars from './dev'

const FillBidTokenModal = props => {
  const initialState = {
    isClosed: true,
    amount: devVars && devVars.fill.amountBid || '',
    success: null,
    txid: null,
    error: null,
  };
  const [state, setState] = useState(initialState);

  const updateInput = e => {
    if (e.target.name === 'amount') {
      e.target.value = e.target.value.replace(/[^0-9]/g, '');
    }

    setState(prevState => ({
      ...prevState,
      [e.target.name]: e.target.value,
      error: null,
      success: null,
      txid: null,
    }));
  }

  useEffect(() => {
    writeLog('fillbid token modal state', state);
  });

  const close = () => {
    setState({
      ...initialState,
      isClosed: true
    });
  }

  const open = () => {
    setState({
      ...initialState,
      isClosed: false
    });
  }

  const sellToken = async () => {
    const {chain, address, order} = props;
    
    if (Number(state.amount) > order.askamount ||
        Number(state.amount) < 1) {
      setState(prevState => ({
        ...prevState,
        success: null,
        txid: null,
        error: order.askamount === 1 ? 'Amount must be equal to 1' : 'Amount must be between 1 and ' + order.askamount,
      }));
    } else if (toSats(order.price * state.amount) > getMaxSpendNormalUtxos(props.normalUtxos)) {
      setState(prevState => ({
        ...prevState,
        success: null,
        txid: null,
        error: 'Not enough balance',
      }));
    } else {
      try {
        let inputsData, rawtx;
        writeLog('fillbidToken');
        writeLog(TokensLib.V2Assets);

        inputsData = {
          transactionsMany: await Blockchain.tokenTransactionsMany(order.tokenid, order.txid),
          ccUtxos: await Blockchain.addCCInputs(order.tokenid, address.pubkey, Number(state.amount)),
          normalUtxos: await Blockchain.createCCTx(toSats(order.price * state.amount) + 10000, address.pubkey),
        };

        writeLog('fillbidToken inputs data', inputsData);

        try {
          rawtx = await TokensLib.V2Assets.buildTokenv2fillbid(
            order.tokenid,
            order.txid,
            Number(state.amount),
            order.price,
            props.wif,
            inputsData,
          );
        } catch (e) {
          writeLog(e);
          setState(prevState => ({
            ...prevState,
            success: null,
            txid: null,
            error: e.message,
          }));
        }
    
        writeLog('fillbidToken token rawtx', rawtx);
    
        if (rawtx && rawtx.substr(0, 2) === '04') {
          const {txid} = await Blockchain.broadcast(rawtx);
    
          if (!txid || txid.length !== 64) {
            setState(prevState => ({
              ...prevState,
              success: null,
              txid: null,
              error: 'Unable to broadcast transaction!',
            }));
          } else {
            setState(prevState => ({
              ...prevState,
              success: `${chains[chain].explorerUrl}/${order.tokenid}/transactions/${txid}/${chain}`,
              txid,
              error: null,
              price: '',
              token: null,
              amount: '',
              tokenDropdownOpen: false,
            }));
            //props.setActiveToken();
            setTimeout(() => {
              props.syncData();
            }, 100);
          }
        } else {
          setState(prevState => ({
            ...prevState,
            success: null,
            txid: null,
            error: 'Unable to build transaction!',
          }));
        }
      } catch (e) {
        setState(prevState => ({
          ...prevState,
          success: null,
          txid: null,
          error: e.message,
        }));
      }
    }
  }

  const getTokenData = tokenid => {
    const tokenInfo = props.tokenList.filter(tokenInfo => tokenInfo.tokenid === tokenid)[0];
    return tokenInfo;
  }

  const render = () => {
    const {order} = props;
    const tokenInfo = getTokenData(order.tokenid);
    const maxNormalSpendValue = getMaxSpendNormalUtxos(props.normalUtxos);

    return (
      <React.Fragment>
        <div
          className={`fill-buy-order-trigger ${maxNormalSpendValue === 0 ? ' disabled' : ''}`}
          onClick={() => open()}>
          <i className="fa fa-dollar-sign"></i>
          Sell
        </div>
        <Modal
          show={state.isClosed === false}
          handleClose={() => close()}
          isCloseable={true}
          className="Modal-send-token">
          <div className="create-token-form">
            <h4>Sell token</h4>
            <p>Fill out the form below</p>
            <div className="input-form">
              <input
                type="text"
                name="token"
                value={`Token: ${tokenInfo.name}`}
                disabled
                className="form-input" />
              <input
                type="text"
                name="amount"
                placeholder="Amount (qty)"
                value={state.amount}
                onChange={updateInput}
                className="form-input" />
              <input
                type="text"
                name="token"
                value={`Price: ${order.price}`}
                disabled
                className="form-input" />
              <input
                type="text"
                name="token"
                value={`Total: ${(state.amount || 0) * order.price}`}
                disabled
                className="form-input" />
              <button
                type="button"
                onClick={sellToken}
                disabled={
                  !state.amount ||
                  maxNormalSpendValue === 0
                }
                className="form-input">Sell</button>
              {state.success &&
                <div className="success">
                  Token sell order filled!
                  <div className="txid-label">
                    <strong>Transaction ID:</strong> {state.txid}
                  </div>
                  <a
                    href={state.success}
                    target="_blank">Open on explorer</a>
                </div>
              }
              {state.error &&
                <div className="error">
                  <div>
                    <strong>Error!</strong>
                    <div>{state.error}</div>
                  </div>
                </div>
              }
            </div>
          </div>
        </Modal>
      </React.Fragment>
    );
  }

  return render();
}

export default FillBidTokenModal;