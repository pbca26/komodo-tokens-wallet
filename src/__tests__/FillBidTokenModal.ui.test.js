import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent, screen, waitFor} from '@testing-library/react';
import FillBidTokenModal from '../FillBidTokenModal';
import '../__mocks__/ccimport';
import blockchainMockData from '../__mocks__/blockchain.json';
import {chains} from '../constants';
import 'regenerator-runtime/runtime';

import mockFetch from '../__mocks__/fetch';
import blockchain from '../blockchain';

global.fetch = mockFetch;
blockchain.setExplorerUrl('');

// TODO: (1) test keying input data actually changes state
//       (2) erroneous input

test('FillBidTokenModal component render and token fill bid order event check', async () => {
  const mProps = {
    tokenList: blockchainMockData.tokens.tokens,
    order: {
      'funcid': 'b',
      'txid': 'f196eefba29cfa1e8636ac97d0bd51196b062b4757107ef8e192d2164b1c8b06',
      'bidamount': 0.00001,
      'origaddress': 'CHpN2oZ8oCdjR8sS85NEZEDe9nrVYwQP6E',
      'origtokenaddress': 'CaopajuemreFwEAfbywvFi8oFyrJPNkDs1',
      'tokenid': '7e2b623cb57b44b8dc5d3a3cc36c20d08f66023a85fa2a4490a8fb8783d38347',
      'totalrequired': 1,
      'price': 0.00001,
      'ExpiryHeight': 209614
    },
    normalUtxos: blockchainMockData.utxo,
    wif: 'UtaKVKuNvDVYxqS8PhRESwKeRYHhzYYkEGgvwEJ9XvghuRVs1VNM',
    address: {
      normal: 'RRF6ocq94kLpaH6wPdZjTqQa9mAmT1GToc',
      cc: 'CaopajuemreFwEAfbywvFi8oFyrJPNkDs1',
      pubkey: '03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3'
    },
    chain: 'TKLTEST',
  };

  //console.warn('mProps', mProps)

  const component = renderer.create(
    <FillBidTokenModal {...mProps} />,
  );
  
  let treeJSON = component.toJSON();
  expect(treeJSON).toMatchSnapshot();

  let tree = component.toTree();

  console.log(tree.instance.state)
  const {getAllByText, getByText, getByRole, getAllByRole, getByPlaceholderText, getByDisplayValue} = render(<FillBidTokenModal {...mProps} />);

  // check if all labels/titles/form elements are present
  expect(getAllByRole('button').length).toBe(2);
  expect(getAllByRole('textbox').length).toBe(4);
  
  // form name
  expect(getByText('Sell token')).toBeDefined();
  expect(getByText('Fill out the form below')).toBeDefined();
  // submit button
  expect(getAllByText('Sell')).toBeDefined();
  
  // token name
  expect(getByDisplayValue('Token: NFTJS')).toBeDefined();
  // token fill amount
  expect(getByPlaceholderText('Amount (qty)')).toBeDefined();
  // token order price per unit
  expect(getByDisplayValue('Price: 0.00001')).toBeDefined();
  // token order total price
  expect(getByDisplayValue('Total: 0.00001')).toBeDefined();
  
  // trigger token fill bid event
  fireEvent.click(screen.getAllByText('Sell')[1]);

  await waitFor(() => getByText('Token sell order filled!'));

  expect(getByText('Token sell order filled!')).toBeDefined();
  expect(getByText('Transaction ID:')).toBeDefined();
  // txid
  expect(getByText('1111111111111111111111111111111111111111111111111111111111111111')).toBeDefined();
  expect(getByText('Open on explorer')).toBeDefined();  
  expect(getByRole('link', {href: 'http://explorer.komodoplatform.com:20000/tokens/7e2b623cb57b44b8dc5d3a3cc36c20d08f66023a85fa2a4490a8fb8783d38347/transactions/1111111111111111111111111111111111111111111111111111111111111111/TKLTEST'})).toBeDefined();
});