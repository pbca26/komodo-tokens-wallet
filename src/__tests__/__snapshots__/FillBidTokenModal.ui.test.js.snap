// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FillBidTokenModal component render and token fill bid order event check 1`] = `
Array [
  <div
    className="fill-buy-order-trigger "
    onClick={[Function]}
  >
    <i
      className="fa fa-dollar-sign"
    />
    Sell
  </div>,
  <div
    className="Modal modal  Modal-send-token"
  >
    <div
      onClick={[Function]}
    >
      <div
        className="modal-background"
      />
    </div>
    <div
      className="modal-content"
    >
      <div
        className="card"
      >
        <button
          className="modal-close is-large "
          onClick={[Function]}
        />
        <div
          className="card-content"
        >
          <div
            className="content"
          >
            <div
              className="create-token-form"
            >
              <h4>
                Sell token
              </h4>
              <p>
                Fill out the form below
              </p>
              <div
                className="input-form"
              >
                <input
                  className="form-input"
                  disabled={true}
                  name="token"
                  type="text"
                  value="Token: NFTJS"
                />
                <input
                  className="form-input"
                  name="amount"
                  onChange={[Function]}
                  placeholder="Amount (qty)"
                  type="text"
                  value={1}
                />
                <input
                  className="form-input"
                  disabled={true}
                  name="token"
                  type="text"
                  value="Price: 0.00001"
                />
                <input
                  className="form-input"
                  disabled={true}
                  name="token"
                  type="text"
                  value="Total: 0.00001"
                />
                <button
                  className="form-input"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Sell
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
